packages/manager/modules/pci/src/projects/project/storages/databases/database/service-integration/add/add.controller.js

```diff
@@ -6,47 +9,196 @@ export default class {
+          ? [this.database]
+          : this.servicesList.filter((db) =>
+              destinationEngines.includes(db.engine),
+            );
+    }
+    // Preselect service if only one element
+    this.sourceService =
+      this.sourceServiceList.length === 1 ? this.sourceServiceList[0] : null;
+    this.destinationService =
+      this.destinationServiceList.length === 1
+        ? this.destinationServiceList[0]
+        : null;
+    // Emtpy parameters array
+    this.parameters = {};
+
+    // Check if preselected data is valid
+    this.checkIntegrationValidity();
+  }
+
+  get isDisabledSourceSelect() {
+    if (!this.sourceServiceList || this.sourceServiceList.length === 0) {
+      return true;
+    }
+    return (
+      this.integrationCapability.sourceEngines.includes(this.database.engine) &&
+      !this.integrationCapability.destinationEngines.includes(
+        this.database.engine,
+      )
+    );
+  }
+
+  get isDisabledDestinationSelect() {
+    if (
+      !this.destinationServiceList ||
+      this.destinationServiceList.length === 0
+    ) {
+      return true;
+    }
+    return (
+      this.integrationCapability.destinationEngines.includes(
+        this.database.engine,
+      ) &&
+      !this.integrationCapability.sourceEngines.includes(this.database.engine)
+    );
+  }
+
+  checkIntegrationValidity() {
+    // We can not add twice the same integration
+    this.existingIntegrationError =
+      this.serviceIntegrationList.filter(
+        (existingIntegration) =>
+          existingIntegration.type === this.integrationCapability?.type &&
+          existingIntegration.sourceServiceId === this.sourceService?.id &&
+          existingIntegration.destinationServiceId ===
+            this.destinationService?.id,
+      ).length > 0;
+    // We have to select the current service as either source or destination
+    this.notCurrentServiceSelectedError =
+      this.sourceService?.id !== this.database.id &&
+      this.destinationService?.id !== this.database.id;
+    // Show a message if no service matches the selected integration type
+    this.showNoServiceMessage =
+      this.sourceServiceList.length === 0 ||
+      this.destinationServiceList.length === 0;
+    // List of expected engines to add for current integration type
+    this.expectedEngines =
+      this.sourceServiceList.length === 0
+        ? this.integrationCapability.sourceEngines
+        : this.integrationCapability.destinationEngines;
+  }
+
+  formatParameters() {
+    // Do not post parameter property at all if no parameter
+    if (Object.keys(this.parameters).length === 0) {
+      return undefined;
+    }
+    // API only accepts string values
+    const formatedParameters = {};
+    Object.keys(this.parameters).forEach((parameterName) => {
+      formatedParameters[parameterName] = this.parameters[
+        parameterName
+      ].toString();
+    });
+    return formatedParameters;
+  }
```

Comment:
**suggestion:** Can be combined into a single statement using `Array.reduce`

Reply:
It seems that `Array.reduce` is less performant than `forEach`